### PR TITLE
add satrn

### PR DESCRIPTION
**Pull request by zhiminzhang0830**
复现论文：On Recognizing Texts of Arbitrary Shapes with 2D Self-Attention
参考代码：https://github.com/open-mmlab/mmocr/blob/1.x/configs/textrecog/satrn/README.md

### CODE REVIEW

**Comment by zhiminzhang0830** [comment]
数据集：https://aistudio.baidu.com/aistudio/datasetdetail/166485
实验结果：
IIIK-3000:94.53，SVT：91.04，IC13:94.68，IC15:78.24，SVTP：83.72，CUTE80:86.11，Avg:88.05

模型训练：
python3 -m paddle.distributed.launch --log_dir=./debug/ --gpus '0,1,2,3'  tools/train.py -c configs/rec/rec_satrn.yml

模型验证：
python tools/eval.py -c {your config file} \
                    -o Global.pretrained_model={your model file} \
                    Eval.dataset.data_dir={your dataset path}/IIIT5k_3000

模型测试：
python3 tools/infer_rec.py -c {your config file} \
        -o Global.pretrained_model={your model file}  \
        Global.infer_img="doc/imgs_words_en/"

存在问题：
1.使用python tools/export_model.py导出模型时，速度非常慢，大概需要7分钟才能导出模型
2.使用导出的模型做推理的时候报错，报错信息如下：
[libprotobuf ERROR /paddle/build/third_party/protobuf/src/extern_protobuf/src/google/protobuf/io/coded_stream.cc:208] A protocol message was rejected because it was too big (more than 67108864 bytes).  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
[libprotobuf ERROR /paddle/build/third_party/protobuf/src/extern_protobuf/src/google/protobuf/io/coded_stream.cc:208] A protocol message was rejected because it was too big (more than 67108864 bytes).  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
Traceback (most recent call last):
  File "tools/infer/predict_rec.py", line 690, in <module>
    main(utility.parse_args())
  File "tools/infer/predict_rec.py", line 652, in main
    text_recognizer = TextRecognizer(args)
  File "tools/infer/predict_rec.py", line 127, in __init__
    utility.create_predictor(args, 'rec', logger)
  File "/data/code/PaddleOCR_satrn/tools/infer/utility.py", line 277, in create_predictor
    predictor = inference.create_predictor(config)
ValueError: (InvalidArgument) Failed to parse program_desc from binary string.
  [Hint: Expected desc_.ParseFromString(binary_str) == true, but received desc_.ParseFromString(binary_str):0 != true:1.] (at /paddle/paddle/fluid/framework/program_desc.cc:103)

**Comment by zhiminzhang0830** [comment]
模型链接：https://pan.baidu.com/s/10J-Bsd881bimKaclKszlaQ?pwd=lk8a

**Comment by zhiminzhang0830** [comment]
更正一下：

**Comment by Topdu** [comment]
satrn_head.py的526行修改为：for step in range(0, paddle.to_tensor(self.max_seq_len)):
这样可以解决导出inference model 慢的问题呢，但是推理的结果是：
Predicts of ./doc/imgs_words_en/word_19.png:('slowuknuknuknuknuknuknuknuknuknuknukniuknuknuknuknuknuknuknuknukn', 0.5304282307624817)
看结果分析，后处理似乎没有找到eos。
推理命令：
python tools/infer/predict_rec.py --image_dir='./doc/imgs_words_en/word_19.png' --rec_model_dir='./inference/satrn/' --rec_algorithm='SATRN' --rec_image_shape='3,32,100' --rec_char_dict_path='./ppocr/utils/dict90.txt'

**Comment by Topdu** [comment]
satrn 和 nrtr是同类型的识别模型，如果可以的话尽量复用nrtr的代码，例如shallow cnn可以写到MTB中，attention和encoder layer和decoder layer 如果差别不大的话也可以复用nrtr的代码

File: ppocr/modeling/heads/rec_satrn_head.py [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+
+        src_mask = self._get_mask(out_enc, valid_ratio)
+        N = out_enc.shape[0]
+        init_target_seq = paddle.full(
+            (N, self.max_seq_len + 1), self.padding_idx, dtype='int64')
+        # bsz * seq_len
+        init_target_seq[:, 0] = self.start_idx
+
+        outputs = []
+        for step in range(0, self.max_seq_len):
```
Comment (Topdu) [inline ppocr/modeling/heads/rec_satrn_head.py:526]: for step in range(0, paddle.to_tensor(self.max_seq_len)):

**Comment by zhiminzhang0830** [comment]
导出模型：

File: configs/rec/rec_satrn.yml [hunk 2]
```diff
@@ -0,0 +1,117 @@
+Global:
+  use_gpu: true
+  epoch_num: 5
+  log_smooth_window: 20
+  print_batch_step: 50
+  save_model_dir: ../work_dir/ppocr/satrn_branch/rec_satrn/
```
Comment (Topdu) [inline configs/rec/rec_satrn.yml:6]: save_model_dir最好与其他方法修改一致：./output/rec/rec_satrn

File: configs/rec/rec_satrn.yml [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  name: SATRNLabelDecode
+
+Metric:
+  name: RecMetric
+  main_indicator: acc
+
+Train:
+  dataset:
+    name: LMDBDataSet
+    data_dir: /data/Dataset/OCR_Rec/visual_data/rfl_dataset2/training
```
Comment (Topdu) [inline configs/rec/rec_satrn.yml:78]: ./train_data/data_lmdb_release/training/

File: configs/rec/rec_satrn.yml [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    shuffle: True
+    batch_size_per_card: 128
+    drop_last: True
+    num_workers: 8
+    use_shared_memory: False
+
+Eval:
+  dataset:
+    name: LMDBDataSet
+    data_dir: /data/Dataset/OCR_Rec/visual_data/rfl_dataset2/evaluation_academic
```
Comment (Topdu) [inline configs/rec/rec_satrn.yml:99]: ./train_data/data_lmdb_release/evaluation/

File: ppocr/data/imaug/rec_img_aug.py [hunk 5]
```diff
@@ -465,6 +465,21 @@ def __call__(self, data):
         return data
 
 
+class SATRNRecResizeImg(object):
+    def __init__(self, image_shape, padding=True, **kwargs):
```
Comment (Topdu) [inline ppocr/data/imaug/rec_img_aug.py:469]: 如果没用到SATRNRecResizeImg的话可以删除

File: ppocr/modeling/heads/rec_satrn_head.py [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+    def forward(self, q, k, v, mask=None):
+        def masked_fill(x, mask, value):
+            y = paddle.full(x.shape, value, x.dtype)
+            return paddle.where(mask, y, x)
+
+        attn = paddle.matmul(q / self.temperature, k.transpose([0, 1, 3, 2]))
+        if mask is not None:
+            attn = masked_fill(attn, mask == 0, -1e9)
+            # attn = attn.masked_fill(mask == 0, float('-inf'))
+            # attn += mask
```
Comment (tink2123) [inline ppocr/modeling/heads/rec_satrn_head.py:190]: todo：不必要的注释可以删除

**Review by tink2123 — APPROVED** [review]
LGTM

**Comment by tink2123** [comment]
需要补充文档并接入TIPC

**Pull request merged**